src/plugins/unified_histogram/public/container/container.tsx

```diff
@@ -0,0 +1,150 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React, { forwardRef, useEffect, useImperativeHandle, useMemo, useState } from 'react';
+import { Subject } from 'rxjs';
+import { UnifiedHistogramLayout, UnifiedHistogramLayoutProps } from '../layout';
+import type { UnifiedHistogramInputMessage } from '../types';
+import {
+  UnifiedHistogramState,
+  UnifiedHistogramStateOptions,
+  UnifiedHistogramStateService,
+} from './services/state_service';
+import { useStateProps } from './hooks/use_state_props';
+
+type LayoutProps = Pick<
+  UnifiedHistogramLayoutProps,
+  'services' | 'disableAutoFetching' | 'disableTriggers' | 'disabledActions'
+>;
+
+/**
+ * The props exposed by the container
+ */
+export type UnifiedHistogramContainerProps = Pick<
+  UnifiedHistogramLayoutProps,
+  'className' | 'resizeRef' | 'appendHitsCounter' | 'children'
+>;
+
+/**
+ * The options used to initialize the container
+ */
+export type UnifiedHistogramInitializeOptions = UnifiedHistogramStateOptions &
+  Omit<LayoutProps, 'services'>;
+
+/**
+ * The uninitialized API exposed by the container
+ */
+export interface UnifiedHistogramUninitializedApi {
+  /**
+   * Whether the container has been initialized
+   */
+  initialized: false;
+  /**
+   * Initialize the container
+   */
+  initialize: (options: UnifiedHistogramInitializeOptions) => void;
+}
+
+/**
+ * The initialized API exposed by the container
+ */
+export interface UnifiedHistogramInitializedApi {
+  /**
+   * Whether the container has been initialized
+   */
+  initialized: true;
+  /**
+   * Get an observable to watch the container state, optionally providing a
+   * selector function which only gets triggered when the selected state changes
+   */
+  getState$: UnifiedHistogramStateService['getState$'];
+  /**
+   * Update the container state my providing a partial state object
+   */
+  updateState: UnifiedHistogramStateService['updateState'];
+  /**
+   * Manually trigger a refetch of the data
+   */
+  refetch: () => void;
+}
+
+/**
+ * The API exposed by the container
+ */
+export type UnifiedHistogramApi = UnifiedHistogramUninitializedApi | UnifiedHistogramInitializedApi;
+
+export const UnifiedHistogramContainer = forwardRef<
+  UnifiedHistogramApi,
+  UnifiedHistogramContainerProps
+>((containerProps, ref) => {
+  const [initialized, setInitialized] = useState(false);
+  const [layoutProps, setLayoutProps] = useState<LayoutProps>();
+  const [state, setState] = useState<UnifiedHistogramState>();
+  const [stateService, setStateService] = useState<UnifiedHistogramStateService>();
+  const stateProps = useStateProps({ state, stateService });
+  const input$ = useMemo(() => new Subject<UnifiedHistogramInputMessage>(), []);
+
+  useImperativeHandle(
```

Comment:
Can you explain to me what this code is doing? It appears to be doing some initialization but its unclear to me why `useImperativeHandle` is needed and what special work its doing.

Reply:
`useImperativeHandle` allows us to expose the API to the consumer as a ref object, and the `initialize` function is how we receive the required services and initial state for Unified Histogram. I separated the API object to an `api` variable and added a comment to the `useImperativeHandle` call to make it a bit clearer.

Reply:
Its my personal take that this code is working for `useImperativeHandle` rather than `useImperativeHandle` providing value over a plain js equivalent. That said, I understand the Building Blocks project is leaning in this direction. I've said my 2¢ and can work with this.

Reply:
I think this is a totally valid criticism, even if it pushes back on the current ideas around Building Blocks. In your "plain js equivalent", are you imagining something along the lines of the consumer instantiating an API/service/controller directly and passing it down to the Building Block as a shared source of truth (that's what I'm imagining at least 😅)? If so, that would probably work great for Unified Histogram, but might not for other Building Blocks that are more complex or built around embeddables.

Reply:
This is my naive approach - https://jsfiddle.net/mattkime/106b432k/4/ - I'm sure it doesn't account for the complexity that Building Blocks require. I should probably study up on embeddables to better understand the problem.